dvc/utils/__init__.py

```diff
@@ -367,7 +367,7 @@ def resolve_paths(repo, out):
     if (
         not scheme
         and abspath.isin_or_eq(repo.root_dir)
-        and not contains_symlink_up_to(abspath, repo.root_dir)
+        and not contains_symlink_up_to(dirname, repo.root_dir)
```

Comment:
We want to place the `.dvc` files into `cwd()` when the path leading up to the output contains a link, rather than when the output itself is a link, so we only need to check up to `dirname` here rather than `abspath`.

If the output itself is a symlink, it should not matter to DVC, as we will end up copying the file contents into DVC cache, and then replacing the original link with our own link or copy to the DVC cache entry. In that case, it is still safe for us to place the `.dvc` into the same directory as the original symlink.